### DISCUSSION TITLE
Can`t get SML read into Scripting v2=SML#DJ_TPWRIN

### DISCUSSION THREAD

**Discussion by HofeBY** [arendst/Tasmota#17574] [Q&A]
Hi
Can`t get read the v2 Value of SML#DJ_TPWRIN. It´s always zero.
Powermeter get read well in my example.
Maybe the example is outdated.

What do i have to change ?

[Tasmota 12.3.1.2 von Theo Arends](https://bit.ly/tasmota)

My current script:
>D
r="1,770701000"
s="1,77070100100700ff"
qssi=0
CPUtemp=0
hum=0
temp=0
press=0
s1="Tagesverbrauch: {m}<b><span style=>"
v2=0
o2=0
hr=0
md=0
yr=0
;Strom
p:sm=0
sd=0
sma=0
smn=0
sya=0
syn=0
sspr=0.3516
yspr=0
ysgp=99.09
y€=0
 
>T
; get telemetry sensor values
qssi=Wifi#RSSI
CPUtemp=ESP32#Temperature
temp=BME280#Temperature
hum=BME280#Humidity
press=BME280#Pressure

v2=SML#DJ_TPWRIN
;v2=4168.8057
 
>B
->sensor53 r
tper=10
 
>S
;Tagesverbrauch
hr=hours
if chg[hr]>0
and hr==0
and v2>0
then
sm=v2
svars
endif
if upsecs%tper==0
then
sd=v2-sm
endif
 
;Monatsverbrauch
md=day
if chg[md]>0
and md==1
and v2>0
then
sma=v2
svars
endif
if upsecs%tper==0
then
smn=v2-sma
endif 
 
;Jahresverbrauch
yr=year
if chg[yr]>0
and hr==0
and md==1
and v2>0
then
sya=v2
svars
endif
if upsecs%tper==0
then
syn=v2-sya
 
;Kosten
yspr=syn*sspr+ysgp
endif
 
>J
,"Verbrauch Tag":"%sd%"
,"Verbrauch Monat":"%smn%"
,"Verbrauch Jahr":"%syn%"
,"Zählerstand 0:00Uhr":"%4sm%"
,"Zählerstand Monatsanfang":"%4sma%"
,"Zählerstand Jahrenanfang":"%4sya%"
,"Kosten dieses Jahr":"%2yspr%"
 
>W
==============
WLAN Empfang: {m}%qssi%
==============
Ausgabe der Variablenwerte:
%dvnm% hat IP %lip%
Local Timestamp: %tstamp%
Datum %0day%.%0month%.%0year% Uhrzeit %0hours%:%0mins%:%0secs%
Globale Werte: %2gtmp% %2ghum% %2gprs%
Temperatur:      {m} %2temp%
Feuchtigkeit:      {m} %2hum%
Druck:      {m} %2press%
CPU Temperatur:      {m} %1CPUtemp%
==============
Tagesverbrauch %0day%.%0month%.%year%:  {m} %2sd%  KWh
Monatsverbrauch %0month%.%0year%: {m} %2smn% KWh
Jahresverbrauch %0year%: {m} %2syn% KWh
==============
Strom Preis/kWh:       {m} %4sspr% €
Strom Grundpreis:      {m} %2ysgp% €
--------------
Strom Kosten bis %0day%.%0month%.%0year%:{m} %2yspr% €
==============
Strom Zählerstände 1.8.0 Bezug:
1.8.0 akt. Bezug:      {m} %4v2% KWh
1.8.0 0:00 Uhr:        {m} %4sm% KWh
1.8.0 Monatsanfang:    {m} %4sma% KWh
1.8.0 Jahresanfang:    {m} %4sya% KWh
==============

>M 1
; ESP32 DevKitC V4 von AZ-Delivery.com
;erster Zähler +1; RxD GPIO 2;Protokoll SML s; Pullup ein 1; Baudrate 9600; Bezeichnung 1 MT681; TxD GPIO 0; Wartezeit*100ms; Hexstring
+1,2,s,1,9600,1 MT681,0,10,
;+2,4,s,1,9600,2 MT681,5,10,
;+3,12,s,1,9600,3 MT681,14,10,
;+4,15,s,1,9600,4 MT681,13,10,
;
1,=h - SML Zähler -
%r%10800ff@1000,1.8.0 Bezug HT&NT,kWh,Total_in,4
%r%10801ff@1000,1.8.1 Bezug HT,kWh,Total_in_0,4
%r%10802ff@1000,1.8.2 Bezug NT,kWh,Total_in_1,4
%r%20800ff@1000,2.8.0 Lieferung HT&NT,kWh,Total_out,4
%r%20801ff@1000,2.8.1 Lieferung HT,kWh,Total_out_0,4
%r%20802ff@1000,2.8.2 Lieferung NT,kWh,Total_out_1,4
%s%@1,16.7.0 akt.Leistung,W,Power_curr,1
1,=h - CPU Werte -
#

==============================================

And here what i get:

![Tasmota-Ergebnis1](https://user-images.githubusercontent.com/121832819/210284395-e74cace4-4ae1-4028-b513-9bff58993236.jpg)
![Tasmota-Ergebnis2](https://user-images.githubusercontent.com/121832819/210284399-f26e972f-a517-439a-9ec4-bb693b27fa72.jpg)

**Comment by gemu2015** [comment]
look a your meter definition or better observe your MQTT messages which may be read in the >T section
v2=SML#DJ_TPWRIN
will not work 

+1,2,s,1,9600,1 MT681,0,10,
your meter has the MQTT-ID  MT681 not SML

%r%10800ff@1000,1.8.0 Bezug HT&NT,kWh,Total_in,4
your decode line has the MQTT id  Total_in not DJ_TPWRIN

so your >T line should be
v2=MT681#Total_in

**Reply by HofeBY** [reply]
Yesss, yess,yess
OMG
Thats true

Changed Section "+1,2,s,1,9600,1 MT681,0,10,"  to "+1,2,s,1,9600,1-MT681,0,10,"
also changed "v2=MT681#DJ_TPWRIN" to "v2=1-MT681#Total_in", now seen in IO-Broker.
I was to blind to see it.

It´s now much more clear for me, how Tasmota and mqtt works.

Great Job, Tasmota Team.

**Comment by NearMediator** [comment]
Buy 0day/1day IoT RCE Geo USA/Europe Escrow only
Telegram: nocaptar
Tox: 72ADC908C245B6596C115AD9131EEB0C2874CD91F044FA268B19F66A85E15A63DBBA827E15D1
